main/modes/mode_diceroller.c

```diff
@@ -0,0 +1,635 @@
+
+const double pi = 3.14159265359;
+
+void diceEnterMode(display_t* disp);
+void diceExitMode(void);
+void diceMainLoop(int64_t elapsedUs);
+void diceButtonCb(buttonEvt_t* evt);
+//void diceTouchCb(touch_event_t* evt);
+//void diceAccelerometerCb(accel_t* accel);
+//void diceAudioCb(uint16_t* samples, uint32_t sampleCnt);
+coord_t* getRegularPolygonVertices(int8_t sides, float rotDeg, int16_t radius);
+void drawRegularPolygon(int xCenter, int yCenter, int8_t sides, float rotDeg, int16_t radius, paletteColor_t col, int dashWidth);
+void changeActiveSelection();
+void changeDiceCountRequest(int change);
+void changeDiceSidesRequest(int change);
+void doRoll(int count, int sides, int index);
+void doStateMachine(int64_t elapsedUs);
+
+const int MAXDICE = 6;
+const int COUNTCOUNT = 7;
+int validSides[] = {4, 6, 8, 10, 12, 20, 100};
+int polygonSides[] = {3, 4, 3, 4, 5, 3, 6};
+
+int64_t rollAnimationPeriod = 1000000; //1 Second Spin
+int64_t fakeValRerollPeriod = 200000; //200 ms
+double spinScaler = 1;
+
+enum dr_stateVals 
+{
+    DR_STARTUP = 0,
+    DR_SHOWROLL = 1,
+    DR_ROLLING = 2
+};
+
+//int polygonOuterSides[] = []
+//Consider adding outer geometry of dice to make them more recognizable
+
+swadgeMode modeDiceRoller =
+{
+    .modeName = "Dice Roller",
+    .fnEnterMode = diceEnterMode,
+    .fnExitMode = diceExitMode,
+    .fnMainLoop = diceMainLoop,
+    .fnButtonCallback = diceButtonCb,
+    .fnTouchCallback = NULL,
+    .wifiMode = NO_WIFI,
+    .fnEspNowRecvCb = NULL,
+    .fnEspNowSendCb = NULL,
+    .fnAccelerometerCallback = NULL,
+    .fnAudioCallback = NULL,
+    .fnTemperatureCallback = NULL,
+    .overrideUsb = false
+};
+
+typedef struct
+{
+    display_t* disp;
+    font_t ibm_vga8;
+
+    double timeAngle;
+
+    int stateAdvanceFlag;
+    int state;
+
+    int requestCount;
+    int requestSides;
+    int sideIndex;
+
+    //esp_timer_handle_t rollTimer;
+    int64_t rollStartTimeUs;
+    int fakeVal;
+    int fakeValIndex;
+
+    int activeSelection;
+
+    int rollIndex;
+    int rollSides;
+    int rollSize;
+    int* rolls;
+
+    uint32_t timeUs;
+    uint32_t lastCallTimeUs;
+    uint32_t rollerNum;
+} diceRoller_t;
+
+diceRoller_t* diceRoller;
+
+void diceEnterMode(display_t* disp)
+{
+    diceRoller = calloc(1, sizeof(diceRoller_t));
+    
+    diceRoller->disp = disp;
+
+    loadFont("ibm_vga8.font", &diceRoller->ibm_vga8);
+
+    diceRoller->timeAngle = 0;
+    diceRoller->rollSize = 0;
+    diceRoller->rollSides = 0;
+
+    //bootloader_random_enable();
+    diceRoller->rollerNum = 0;
+
+    diceRoller->requestCount = 0;
+    diceRoller->requestSides = 0;
+    diceRoller->sideIndex = 0;
+    diceRoller->activeSelection = 0;
+
+    diceRoller->state = DR_STARTUP;
+    diceRoller->stateAdvanceFlag = 0;
+}
+
+void diceExitMode(void)
+{
+    freeFont(&diceRoller->ibm_vga8);
+    free(diceRoller->rolls);

```

Comment:
It feels a little odd to have this freed here, but not allocated in `diceEnterMode`. Are you sure it's allocated? Maybe check that it's not `NULL` first?

Reply:
Initialized pointer to NULL and made free conditional on exit.